plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc

```diff
@@ -0,0 +1,225 @@
+    varying ScreenSample& sample)
+{
+    Ray ray = sample.ray;
+    varying vec3f color = make_vec3f(0.f);
+
+    sample.z = inf;
+    sample.alpha = 0.f;
+
+    traceRay(self->abstract.super.model, ray);
+
+    if (ray.geomID < 0)
+    {
+        // No intersection. Return skybox color
+        color = make_vec3f(skyboxMapping((Renderer*)self, ray,
+                                         self->abstract.bgMaterial)) *
+                DEFAULT_SKY_POWER_ZERO_BOUNCE;
+
+        sample.alpha = 0.f;
+        return color;
+    }
+
+    // else, we've got a hit!
+    DifferentialGeometry dg;
+    postIntersect(self->abstract.super.model, dg, ray,
+                  DG_NS | DG_NG | DG_NORMALIZE | DG_FACEFORWARD |
+                      DG_MATERIALID | DG_COLOR);
+
+    uniform Material* material = dg.material;
+    uniform ExtendedOBJMaterial* objMaterial =
+        (uniform ExtendedOBJMaterial*)material;
+
+    vec3f Kd = make_vec3f(0.f);
+    if (!objMaterial)
+        Kd = make_vec3f(dg.color);
+    else
+        foreach_unique(mat in objMaterial) Kd = mat->Kd * make_vec3f(dg.color);
+
+    Ray localray = ray;
+    vec3f accucolor = make_vec3f(0.f);
+    vec3f mask = make_vec3f(1.f);
+
+    varying RandomTEA* uniform rng =
+        createRandomNumberGenerator(&self->abstract.super, sample);
+
+    // path tracing loop
+    for (int bounces = 0; bounces < NB_MAX_PATH_TRACING_REBOUNDS; bounces++)
+    {
+        traceRay(self->abstract.super.model, localray);
+
+        // if ray misses scene (no hit occurs), return background colour
+        if (localray.geomID < 0)
+        {
+            const vec3f bgcol =
+                make_vec3f(skyboxMapping((Renderer*)self, localray,
+                                         self->abstract.bgMaterial)) *
+                DEFAULT_SKY_POWER;
+
+            accucolor = accucolor + mask * bgcol;
+            break;
+        }
+        DifferentialGeometry dg;
+        postIntersect(self->abstract.super.model, dg, localray,
+                      DG_NG | DG_NS | DG_NORMALIZE | DG_FACEFORWARD |
+                          DG_MATERIALID | DG_COLOR);
+
+        uniform Material* material = dg.material;
+        uniform ExtendedOBJMaterial* objMaterial =
+            (uniform ExtendedOBJMaterial*)material;
+        if (!objMaterial)
+            Kd = make_vec3f(dg.color);
+        else
+            foreach_unique(mat in objMaterial) Kd =
+                mat->Kd * make_vec3f(dg.color);
+
+        // compute ray direction of uniformly random diffuse ray
+        const vec3f raydir =
+            getRandomVector(sample, dg.Ns, self->abstract.super.epsilon);
+
+        // origin of new ray in path is hitpoint of previous ray in path
+        const vec3f hitpoint = dg.P + self->abstract.super.epsilon * dg.Ns;
+        setRay(localray, hitpoint, raydir);
+        localray.t0 = self->abstract.super.epsilon;
+        localray.t = self->abstract.ambientOcclusionDistance;
```

Comment:
I'm not sure if limiting t (the maximum allowed hit distance along ray) to the ambientOcclusionDistance will work the same as setting it to a very large number. This may or may not result in some weird lighting, but may also be faster.

Reply:
I understand, but the ambient occlusion distance can be set to a very large value anyway. And this way, we can change it at runtime.